### ISSUE TITLE
Bug: Unnecessarily disposing past-child nodes

### ISSUE THREAD

**Issue by losnir** [knockout/knockout#1740]
Here is the scenario:
1. We have a custom KO component that moves a certain DOM element (call it **A**) to the exact point where the component lives in the DOM.
2. In the tear-down sequence of this component, we do not want to let **A** be removed (since it's now contained in the component's element, it will be removed with it).
3. In order to prevent **A**'s deletion, we move it back to it's original place.

However, KO still removes the element, since when the disposal sequence started, it was a child of the disposal subject.

Here is the relevant code snippet -- you can see that before the loops starts, the node children's are being computed. Here lies the problem, since progressing through the loop (and the disposal callbacks) might change the node's hierarchy:

```
cleanNode : function(node) {
                        // First clean this node, where applicable
                        if (cleanableNodeTypes[node.nodeType]) {
                            cleanSingleNode(node);

                            // ... then its descendants, where applicable
                            if (cleanableNodeTypesWithDescendants[node.nodeType]) {
                                // Clone the descendants list in case it changes during iteration
                                var descendants = [];
                                ko.utils.arrayPushAll(descendants, node.getElementsByTagName("*"));
                                for (var i = 0, j = descendants.length; i < j; i++)
                                    cleanSingleNode(descendants[i]);
                            }
                        }
                        return node;
                    },
```

My solution was to protect the call to `cleanSingleNode`:

```
ko.utils.domNodeIsContainedBy(descendants[i], node) && cleanSingleNode(descendants[i]);
```

Thoughts?

**Comment by losnir** [comment]
** Crickets ***

**Comment by mbest** [comment]
Your logic here does make sense, but it does seem like quite an edge case. I'd be concerned about the possible performance penalty of having to check each node before cleaning it for a situation that won't affect most people.

**Comment by SteveSanderson** [comment]
I agree with @mbest.

If this becomes a common feature request, I'm sure we can find a way to handle it efficiently enough, but in the meantime, @losnir would it work for you to make sure your element avoids the cull? Perhaps you could move it to a safe place *before* the dispose starts.

**Comment by losnir** [comment]
Thanks for the replies @mbest and @SteveSanderson.

The sole purpose of the destroy callback is to make sure these elements avoid the cull, but as I've demonstrated -- it is not possible.

Unfortunately we have no control over the disposal of the parent element(s).

Maybe perhaps we can add a "beforeDispose" callback? One that might fire before calculation of child nodes.

**Comment by losnir** [comment]
Any updates on this?

**Comment by idanhen** [comment]
Hi, 

I have the same issue. 
"beforeDispose" callback would be my preferred solution.

**Comment by losnir** [comment]
We have encountered this issue once again today.

@mbest, @SteveSanderson -- I would like your opinions regarding the suggested solution.
 I think the latter can be implemented without a significant cost in performance (as opposed to the former).

**Comment by losnir** [comment]
Hi,

I'm no longer responsible for that project, so this issue does not relate to me directly anymore, however I'd still like to see this resolved and prevent this from happening again to other users (honestly this scenario doesn't seem far fetched to me, looks like a normal use-case for a disposal callback). 

@mbest @SteveSanderson Would you consider a PR for a "beforeDisposal" callback? It's low effort & low risk as far as I can tell.

Nir.

**Comment by mstarets** [comment]
It seems like the same issue still exists for bindings on virtual elements. TConsider the following example:

<!— ko managesChildren —><!— ko text: foo —>some text<!— /ko—><!— /ko—>

When cleanSingleNode processes element, it will call cleanImmediateCommentTypeChildren():

function cleanImmediateCommentTypeChildren(nodeWithChildren) {
var child, nextChild = nodeWithChildren.firstChild;
while (child = nextChild) {
nextChild = child.nextSibling;
if (child.nodeType === 8)
cleanSingleNode(child);
}
}

Now suppose that the 'managesChildren' binding moves its child nodes somewhere to prevent them from being cleaned, and <!— ko text: foo —> got moved. Since nextChild is already assigned the value of child.nextSibling, the <!— ko text: foo —> node will still be cleaned, and the subsequent child.nextSibling will not be pointing at the expected sibling.

**Comment by mbest** [comment]
Fixed the comment-node issue now.

**Comment by mstarets** [comment]
Michael, we'd like to report an issue with this fix that could be affecting applications using Knockout and jQuery/jQueryUI together.

After running our tests with 3.5 beta, we started getting the "ko.cleanNode: An already cleaned node was removed from the document" error. The error was thrown because the DOM subtree contained jQueryUI components modifying their DOM during cleanup. Consider the following case: 
1) A jQueryUI widget is written to call .wrap() at construction time (creating an element wrapper around the initialization element), and to call .unwrap() in its .destroy() method. The restoring of the DOM in destroy() is common practice for all jQueryUI widgets.
2) Knockout calls cleanExternalData() during .cleanNode(), that calls jQuery's cleanData() resulting in the destroy() call on the jQueryUI widget associated with the DOM element being cleaned by Knockout. We'll refer to that element as 'X' below.
3). The destroy() calls .unwrap() to remove the wrapper element around the element 'X'. That replaces the parent element of 'X' with 'X' itself, thus reducing the index of  'X' in the life collection by 1.
4) The check in .cleanNode() throws the "already cleaned" error.

We worked around this problem by short-circuiting .unwrap() calls when the .destroy() method is ultimately invoked by cleanExternalData(). However, we are a bit concerned that other Knockout users may run into similar problems.

We understand that this is 'as-designed' behavior: the code is meant to ensure that no elements in the life collection were pushed back before the current index and thus missed by the cleanup process. We suspect, however, that many Knockout users will prefer the code to be a bit more forgiving and allow self-recovery. Perhaps the code code walk back within the life collection until the missed element is found. We probably would not walk forever for performance reasons, so ko.utils.domNodeDisposal would have a settable constant for the number of elements that should be scanned in that case. The default could be 0 (current behavior with no recovery). In our case, we know that our jQueryUI widgets will not produce an index delta greater than 2 during cleanup, so we would set it to 2.

**Comment by mbest** [comment]
@mstarets Can you provide an reproduction of this problem?

**Comment by mstarets** [comment]
@mbest I will put a small test case together by tomorrow. Thanks!

**Comment by mstarets** [comment]
[ko-clean.zip](https://github.com/knockout/knockout/files/1639682/ko-clean.zip)

Please see the attached test case. Press the 'Clean Subtree' button to see the error. Put a breakpoint in _destroy (index.html) to verify that the widget's _destroy() method is ultimately invoked by cleanExternalData()

**Comment by mbest** [comment]
The changes here were meant to allows users to remove nodes to prevent them from being cleaned, but as @mstarets has pointed out, now prevents already cleaned nodes from being removed. Previously the latter behavior was fine and produced no problems.

I like the idea of walking back through the list of nodes to restart the cleanup process when a node is removed. I'll work on the code changes soon.

**Comment by mbest** [comment]
I've uploaded a fix: #2341
